_blog/kubernetes-chaos-engineering-netsplit/index.md

```diff
@@ -0,0 +1,315 @@
+
+Let's go ahead and delete the routing rules.
+
+In a separate shell, you should monitor the application for time and dropped requests. You could use a command like this:
+
+```bash
+$ while sleep 1; do date +%s; curl -sS http://<your cluster>/ | grep ^Hello; done
+```
+
+In this case, you have the timestamp in the first column and the response from the pod in the other:
+
+```bash
+1519526805 Hello world! via k8s-hello-world-55f48f8c94-vrkr9
+1519526807 Hello world! via k8s-hello-world-55f48f8c94-tjg4n
+```
+
+It's time to drop the bomb. Let's delete the routing rules from the node.
+
+`kube-proxy` can operate in three modes: **userspace**, **iptables** and **ipvs**. The default since Kubernetes 1.2 is **iptables**.
+
+In **iptables** mode, `kube-proxy` writes the list of routing rules to the node using iptables rules.
+
+So you could log in into one of the node servers and delete the iptables rules with `iptables -F`.
+
+If everything went according to plan you should experience something similar to this:
+
+```bash
+1519526805 Hello world! via k8s-hello-world-55f48f8c94-xzvlc
+1519526807 Hello world! via k8s-hello-world-55f48f8c94-tjg4n
+# this is when `iptables -F` was issued
+1519526834 Hello world! via k8s-hello-world-55f48f8c94-vrkr9
+1519526835 Hello world! via k8s-hello-world-55f48f8c94-vrkr9
+```
+
+As you noticed, it took about 27 seconds from when you dropped the iptables rules and the next response, from 1519526807 to 1519526834.
+
+*What happened in this 27 seconds?*
+
+*Why is everything back to normal after 27 seconds?*
+
+Perhaps it's just a coincidence. Let's flush the rules again:
+
+```bash
+1519526855 Hello world! via k8s-hello-world-55f48f8c94-xzvlc
+1519526856 Hello world! via k8s-hello-world-55f48f8c94-tjg4n
+# this is when `iptables -F` was issued
+1519526885 Hello world! via k8s-hello-world-55f48f8c94-npkn6
+1519526887 Hello world! via k8s-hello-world-55f48f8c94-vrkr9
+```
+
+There was a gap of 29 seconds, from 1519526856 to 1519526885, but the cluster is back to normal.
+
+*Why does it take about 30 seconds to reply?*
+
+*Is the node receiving traffic despite no routing table?*
+
+Maybe you could investigate what happens to the node in this 30 seconds. In another terminal, let's monitor requests to the node with:
+
+```bash
+$ while sleep 1; printf %"s\n" $(curl -sS http://<ip of the node>:30000); done
+```
+
+And let's drop the iptables rules. The log from the previous command is:
+
+```bash
+Hello world! via k8s-hello-world-55f48f8c94-xzvlc
+Hello world! via k8s-hello-world-55f48f8c94-tjg4n
+curl: (28) Connection timed out after 10003 milliseconds
+curl: (28) Connection timed out after 10004 milliseconds
+Hello world! via k8s-hello-world-55f48f8c94-npkn6
+Hello world! via k8s-hello-world-55f48f8c94-vrkr9
+```
+
+It shouldn't come as a surprise that connections to the node are timing out after you drop the iptables rules. What's more interesting is that `curls` waits for ten seconds before giving up.
```

Comment:
why not run curl with a shorter timeout?
`--connect-timeout`

Reply:
what's the intent?